src/vstest.console/TestPlatformHelpers/TestRequestManager.cs

```diff
@@ -352,47 +416,116 @@ public void ProcessTestRunAttachments(TestRunAttachmentsProcessingPayload attach
                     EqtTrace.Info("TestRequestManager.ProcessTestRunAttachments: Test run attachments processing completed.");
                     this.testPlatformEventSource.TestRunAttachmentsProcessingRequestStop();
 
-                    // Post the attachments processing complete event
-                    this.metricsPublisher.Result.PublishMetrics(TelemetryDataConstants.TestAttachmentsProcessingCompleteEvent, requestData.MetricsCollection.Metrics);
+                    // Post the attachments processing complete event.
+                    this.metricsPublisher.Result.PublishMetrics(
+                        TelemetryDataConstants.TestAttachmentsProcessingCompleteEvent,
+                        requestData.MetricsCollection.Metrics);
+                }
+            }
+        }
+
+        /// <inheritdoc/>
+        public void StartTestSession(
+            StartTestSessionPayload payload,
+            ITestHostLauncher testHostLauncher,
+            ITestSessionEventsHandler eventsHandler,
+            ProtocolConfig protocolConfig)
+        {
+            EqtTrace.Info("TestRequestManager.StartTestSession: Starting test session.");
+
+            if (payload.TestPlatformOptions != null)
+            {
+                this.telemetryOptedIn = payload.TestPlatformOptions.CollectMetrics;
+            }
+
+            var requestData = this.GetRequestData(protocolConfig);
+
+            if (this.UpdateRunSettingsIfRequired(
+                payload.RunSettings,
+                payload.Sources,
+                null,
+                out string updatedRunsettings))
+            {
+                payload.RunSettings = updatedRunsettings;
+            }
+
+            if (InferRunSettingsHelper.AreRunSettingsCollectorsInCompatibleWithTestSettings(payload.RunSettings))
+            {
+                throw new SettingsException(
+                    string.Format(
+                        Resources.RunsettingsWithDCErrorMessage,
+                        payload.RunSettings));
+            }
+
+            // TODO (copoiena): Collect metrics ?
+
+            lock (this.syncObject)
```

Comment:
it looks like the CreateStartTestSessionRequest isn't necessarily an immediate operation. does this mean that test sessions on our side can only be spawned sequentially? I noticed that there is just a global TestRequestManager instance.

Reply:
That's correct, there's only one instance of the `TestRequestManager` and all operations are sequential, including test discovery and test runs. They're not immediate operations either, meaning we cannot go ahead and execute another request while the current one running. This has been a design limitation for quite some time now and I think it's worth exploring changing this behavior in a PR of its own as I anticipate it'll require significant changes.

However, if we take only test session spawning into consideration this shouldn't be much of an issue as the plan is to spawn the test session(s) during build. Regarding multiple test sessions, is there any real use case for this on the TW side ?

Reply:
our extensibility let's users specify a runsettings file to use separately for each csproj. because of that it's possible that different projects will have a different runsettings

hopefully we'll be good with these starting in sequence and we'll need to be smart about stopping/starting sessions to minimize the cases where the session isn't active at the time we need to do a discovery